File: android/core/src/main/kotlin/dev/aura/core/Workouts.kt
package dev.aura.core

import java.time.LocalDate
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/** One exercise in a template, with its target. */
@Serializable
data class Planned(
    val exercise: String,
    val sets: Int,
    @kotlinx.serialization.SerialName("reps_min") val repsMin: Int,
    @kotlinx.serialization.SerialName("reps_max") val repsMax: Int,
)

/** A starter workout: a named list of exercises with targets. */
@Serializable
data class Template(
    val id: String,
    val name: String,
    val about: String,
    val exercises: List<Planned>,
)

@Serializable
data class CatalogueEntry(
    val name: String,
    val category: String,
    val modality: String,
    val muscles: List<String>,
    val aliases: List<String> = emptyList(),
    @kotlinx.serialization.SerialName("increment_kg") val incrementKg: Double = 2.5,
    @kotlinx.serialization.SerialName("rest_s") val restS: Int = 120,
)

/** Where one exercise of today's plan stands. */
data class PlanEntry(
    /** The exercise as it exists here, which may be an alias of [planned]. */
    val exercise: String,
    /** The template's name for it. */
    val planned: String,
    val sets: Int,
    val repsMin: Int,
    val repsMax: Int,
    /** Working sets of it logged today, from any source. */
    val done: Int,
    /** The most recent working weight, today's included. */
    val lastWeightKg: Double?,
    /** What to aim for today, from the last session before it. */
    val target: Suggestion,
    /** Rest between its sets, in seconds. */
    val restS: Int,
) {
    val complete: Boolean
        get() = done >= sets
}

data class Plan(val template: Template, val day: LocalDate, val entries: List<PlanEntry>) {
    val complete: Boolean
        get() = entries.all { it.complete }
}

/**
 * Starter workouts, and the plan started from one.
 *
 * The templates are generated from the Python app by
 * scripts/export_workouts.py, like the schema, so both apps offer the same
 * sessions and create their exercises with the same muscles. Starting one
 * logs nothing: it makes sure the exercises exist and marks the template as
 * today's plan, under the same settings key the desktop app uses.
 *
 * Exercises are matched by name and by alias, so a template's "Bench Press"
 * is the "Flat Barbell Bench Press" an imported history already has.
 */
object Workouts {

    const val RESOURCE_PATH: String = "/aura/workouts.json"
    const val ACTIVE: String = "workout"

    @Serializable
    private data class Document(
        val catalogue: List<CatalogueEntry>,
        val templates: List<Template>,
        @kotlinx.serialization.SerialName("muscle_targets") val muscleTargets: MuscleTargets,
    )

    private val document: Document by lazy {
        val text =
            Workouts::class.java.getResourceAsStream(RESOURCE_PATH)
                ?.bufferedReader()
                ?.use { it.readText() }
                ?: error(
                    "$RESOURCE_PATH is missing from the classpath. " +
                        "Run scripts/export_workouts.py to generate it."
                )
        Json { ignoreUnknownKeys = true }.decodeFromString(Document.serializer(), text)
    }

    val templates: List<Template>
        get() = document.templates

    val muscleTargets: MuscleTargets
        get() = document.muscleTargets

    fun template(id: String): Template? = templates.firstOrNull { it.id == id }

    private val notAlphanumeric = Regex("[^a-z0-9]+")

    /** "Pull-Up", "pull up" and "Pullup" compare equal, as in workouts.normalise. */
    fun normalise(name: String): String = notAlphanumeric.replace(name.lowercase(), "")

    /** The template's name for an exercise, then its aliases, in preference order. */
    fun namesFor(exercise: String): List<String> {
        val entry = document.catalogue.firstOrNull { it.name.equals(exercise, ignoreCase = true) }
        return listOf(exercise) + entry?.aliases.orEmpty()
    }

    /** An existing exercise, with what counts when choosing between its names. */
    data class Candidate(val id: Int, val name: String, val sets: Int, val imported: Int, val rank: Int)

    // Imported history first, then the most working sets, then the earlier
    // name in the alias list, then the older row. As workouts.Candidate.key.
    private val best: Comparator<Candidate> =
        compareByDescending<Candidate> { it.imported > 0 }
            .thenByDescending { it.sets }
            .thenBy { it.rank }
            .thenBy { it.id }

    private fun existing(db: Db): List<Candidate> =
        db.select(
            """
            SELECT e.id, e.name,
                   COUNT(s.id) AS sets,
                   COUNT(CASE WHEN w.source <> 'manual' THEN s.id END) AS imported
            FROM exercises e
            LEFT JOIN sets s ON s.exercise_id = e.id AND s.is_warmup = 0
            LEFT JOIN workouts w ON w.id = s.workout_id
            GROUP BY e.id, e.name
            """
        ) { Candidate(it.int("id"), it.string("name"), it.int("sets"), it.int("imported"), -1) }

    private fun candidates(existing: List<Candidate>, exercise: String): List<Candidate> {
        val rank = HashMap<String, Int>()
        namesFor(exercise).forEachIndexed { i, name -> rank.putIfAbsent(normalise(name), i) }
        return existing
            .mapNotNull { c -> rank[normalise(c.name)]?.let { c.copy(rank = it) } }
            .sortedWith(best)
    }

    /**
     * The existing exercise a template's exercise means, as (id, name).
     *
     * Any exercise under its name or an alias counts. When several do, the
     * imported one wins, then the one with the most working sets, since that
     * is where the history is. The same rule as workouts.resolve.
     */
    fun resolve(db: Db, exercise: String): Pair<Int, String>? =
        candidates(existing(db), exercise).firstOrNull()?.let { it.id to it.name }

    /** A catalogue exercise found under more than one of its names. */
    data class Merge(val keep: Candidate, val drop: List<Candidate>)

    /** Catalogue exercises that exist under more than one name, and which to keep. */
    fun duplicates(db: Db): List<Merge> {
        val all = existing(db)
        val claimed = HashSet<Int>()
        return document.catalogue.mapNotNull { entry ->
            val found = candidates(all, entry.name).filter { it.id !in claimed }
            if (found.size < 2) return@mapNotNull null
            claimed += found.map { it.id }
            Merge(found.first(), found.drop(1))
        }
    }

    /**
     * Fold each duplicate into the exercise [duplicates] chose to keep: its
     * sets move across and it is deleted. The kept exercise keeps its own
     * name, category and muscles, taking the duplicate's only where it has
     * none. All or nothing, as workouts.merge_duplicates.
     */
    fun mergeDuplicates(db: Db): List<Merge> =
        db.transaction {
            val merges = duplicates(db)
            for (merge in merges) {
                val keep = merge.keep.id
                for (drop in merge.drop.map { it.id }) {
                    db.execute("UPDATE sets SET exercise_id = ? WHERE exercise_id = ?", listOf(keep, drop))
                    db.execute(
                        """
                        INSERT INTO exercise_muscles (exercise_id, muscle, is_primary)
                        SELECT ?, muscle, is_primary FROM exercise_muscles
                        WHERE exercise_id = ?
                          AND NOT EXISTS (SELECT 1 FROM exercise_muscles WHERE exercise_id = ?)
                        """,
                        listOf(keep, drop, keep),
                    )
                    db.execute(
                        """
                        UPDATE exercises
                        SET category = (SELECT category FROM exercises WHERE id = ?)
                        WHERE id = ? AND category IS NULL
                        """,
                        listOf(drop, keep),
                    )
                    db.execute("DELETE FROM exercise_muscles WHERE exercise_id = ?", listOf(drop))
                    db.execute("DELETE FROM exercises WHERE id = ?", listOf(drop))
                }
            }
            merges
        }

    /**
     * Create the template's missing exercises and make it today's plan.
     *
     * An exercise that already exists is left exactly as it is, even with
     * different muscles: it may have come from an import, and its history
     * hangs off it.
     */
    fun start(db: Db, id: String, day: LocalDate): Plan? {
        val chosen = template(id) ?: return null
        val catalogue = document.catalogue.associateBy { it.name.lowercase() }
        for (planned in chosen.exercises) {
            if (resolve(db, planned.exercise) != null) continue
            val entry = catalogue.getValue(planned.exercise.lowercase())
            db.execute(
                "INSERT INTO exercises (name, category, modality) VALUES (?, ?, ?)",
                listOf(entry.name, entry.category, entry.modality),
            )
            val newId =
                exerciseId(db, entry.name) ?: error("${entry.name} was not created")
            for (muscle in entry.muscles) {
                db.execute(
                    "INSERT INTO exercise_muscles (exercise_id, muscle, is_primary) VALUES (?, ?, 1)",
                    listOf(newId, muscle),
                )
            }
        }
        Settings.write(
            db,
            ACTIVE,
            buildJsonObject {
                put("template", id)
                put("day", day.toString())
            },
        )
        return plan(db, day)
    }

    fun finish(db: Db) = Settings.clear(db, ACTIVE)

    /** Today's plan with progress through it, or null if none was started today. */
    fun plan(db: Db, day: LocalDate): Plan? {
        val active = Settings.read(db, ACTIVE) ?: return null
        if (active["day"]?.jsonPrimitive?.content != day.toString()) return null
        val chosen = template(active["template"]?.jsonPrimitive?.content.orEmpty()) ?: return null

        val entries =
            chosen.exercises.map { planned ->
                val found = resolve(db, planned.exercise)
                val id = found?.first
                val loading = document.catalogue.firstOrNull { it.name == planned.exercise }
                val increment = loading?.incrementKg ?: 2.5
                PlanEntry(
                    exercise = found?.second ?: planned.exercise,
                    planned = planned.exercise,
                    sets = planned.sets,
                    repsMin = planned.repsMin,
                    repsMax = planned.repsMax,
                    done = id?.let { doneOn(db, it, day) } ?: 0,
                    lastWeightKg = id?.let { lastWeight(db, it) },
                    // From the last session before today, so it holds steady
                    // through today's sets. As workouts.plan.
                    target =
                        Progression.suggest(
                            id?.let { previousSession(db, it, day) }.orEmpty(),
                            planned.sets,
                            planned.repsMin,
                            planned.repsMax,
                            increment,
                        ),
                    restS = loading?.restS ?: 120,
                )
            }
        return Plan(chosen, day, entries)
    }

    /** Each working set on the last day before [day], in the order done. */
    private fun previousSession(db: Db, exerciseId: Int, day: LocalDate): List<DoneSet> {
        val id = exerciseId.toString()
        val lastDay =
            db.selectOne(
                """
                SELECT MAX(w.performed_on) AS d
                FROM sets s JOIN workouts w ON w.id = s.workout_id
                WHERE s.exercise_id = CAST(? AS INTEGER) AND s.is_warmup = 0 AND w.performed_on < ?
                """,
                listOf(id, day.toString()),
            ) { it.stringOrNull("d") } ?: return emptyList()
        return db.select(
            """
            SELECT s.weight_kg, s.reps, s.rpe
            FROM sets s JOIN workouts w ON w.id = s.workout_id
            WHERE s.exercise_id = CAST(? AS INTEGER) AND s.is_warmup = 0 AND w.performed_on = ?
            ORDER BY w.id, s.position, s.id
            """,
            listOf(id, lastDay),
        ) { DoneSet(it.doubleOrNull("weight_kg"), it.longOrNull("reps")?.toInt(), it.doubleOrNull("rpe")) }
    }

    private fun exerciseId(db: Db, name: String): Int? =
        db.selectOne("SELECT id FROM exercises WHERE lower(name) = lower(?)", listOf(name)) {
            it.int("id")
        }

    private fun doneOn(db: Db, exerciseId: Int, day: LocalDate): Int =
        db.selectOne(
            """
            SELECT COUNT(*) AS n
            FROM sets s JOIN workouts w ON w.id = s.workout_id
            WHERE s.exercise_id = CAST(? AS INTEGER) AND s.is_warmup = 0 AND w.performed_on = ?
            """,
            // Strings only: Android's rawQuery binds every argument as text.
            listOf(exerciseId.toString(), day.toString()),
        ) { it.int("n") } ?: 0

    private fun lastWeight(db: Db, exerciseId: Int): Double? =
        db.selectOne(
            """
            SELECT s.weight_kg
            FROM sets s JOIN workouts w ON w.id = s.workout_id
            WHERE s.exercise_id = CAST(? AS INTEGER) AND s.is_warmup = 0 AND s.weight_kg IS NOT NULL
            ORDER BY w.performed_on DESC, s.id DESC
            LIMIT 1
            """,
            listOf(exerciseId.toString()),
        ) { it.doubleOrNull("weight_kg") }
}

File: api/app/workouts.py
"""Starter workouts: one plain session per body part, to pick from on day one.

A template is a list of exercises with a target of sets and reps. Starting one
does not log anything; it makes sure its exercises exist, with the category
and muscles from the catalogue so the volume views can attribute them, and
marks it as today's plan. The Log screen then shows how far through it you
are and what weight you used last time.

Exercises are matched by name and by alias (`app.seed.ALIASES`), so a
template's "Bench Press" is the "Flat Barbell Bench Press" an imported FitNotes
history already has, and its history carries on.

The templates are generated into the Android app by
`scripts/export_workouts.py`, so both apps offer the same sessions.
"""

from __future__ import annotations

import datetime as dt
import re
from dataclasses import dataclass

from sqlalchemy import and_, case, func, select, text
from sqlalchemy.orm import Session

from app import settings_store
from app.models import Exercise, SetEntry, Workout
from app.seed import ALIASES, CATALOGUE

# The settings key the active plan is kept under. The Android app uses the
# same one, so a restored backup carries it across.
ACTIVE = "workout"

# One session per body part. id, name, what it is for, [(exercise, sets, reps)]
TEMPLATES: list[dict] = [
    {
        "id": "chest",
        "name": "Chest",
        "about": "Flat and incline pressing, then fly and dips.",
        "exercises": [
            ("Bench Press", 4, 8),
            ("Incline Bench Press", 3, 10),
            ("Dumbbell Press", 3, 10),
            ("Chest Fly", 3, 12),
            ("Dip", 3, 10),
        ],
    },
    {
        "id": "back",
        "name": "Back",
        "about": "A hinge, a vertical pull and two rows.",
        "exercises": [
            ("Deadlift", 3, 5),
            ("Pull-Up", 3, 8),
            ("Barbell Row", 3, 8),
            ("Lat Pulldown", 3, 10),
            ("Seated Cable Row", 3, 12),
        ],
    },
    {
        "id": "shoulders",
        "name": "Shoulders",
        "about": "Pressing, then front, side and rear delts.",
        "exercises": [
            ("Overhead Press", 4, 8),
            ("Dumbbell Shoulder Press", 3, 10),
            ("Lateral Raise", 3, 12),
            ("Rear Delt Fly", 3, 15),
            ("Face Pull", 3, 15),
        ],
    },
    {
        "id": "legs",
        "name": "Legs",
        "about": "Quads, hamstrings, glutes and calves.",
        "exercises": [
            ("Back Squat", 4, 6),
            ("Romanian Deadlift", 3, 8),
            ("Leg Press", 3, 10),
            ("Leg Curl", 3, 12),
            ("Calf Raise", 3, 15),
        ],
    },
    {
        "id": "triceps",
        "name": "Triceps",
        "about": "A heavy compound, then all three heads.",
        "exercises": [
            ("Close-Grip Bench Press", 3, 8),
            ("Dip", 3, 10),
            ("Skull Crusher", 3, 10),
            ("Triceps Pushdown", 3, 12),
            ("Overhead Triceps Extension", 3, 12),
        ],
    },
    {
        "id": "biceps",
        "name": "Biceps",
        "about": "Chin-ups, then curls in three grips.",
        "exercises": [
            ("Chin-Up", 3, 8),
            ("Barbell Curl", 3, 10),
            ("Dumbbell Curl", 3, 12),
            ("Hammer Curl", 3, 12),
            ("Preacher Curl", 3, 12),
        ],
    },
]

_CATALOGUE = {name.lower(): (name, category, modality, muscles)
              for name, category, modality, muscles in CATALOGUE}

_NOT_ALPHANUMERIC = re.compile(r"[^a-z0-9]+")


def normalise(name: str) -> str:
    """The form names are compared in: "Pull-Up", "pull up" and "Pullup" agree.

    The Android app normalises the same way, and WorkoutsTest.kt pins it.
    """
    return _NOT_ALPHANUMERIC.sub("", name.lower())


def names_for(exercise: str) -> list[str]:
    """The template's name for an exercise, then its aliases, in preference order."""
    return [exercise, *ALIASES.get(exercise, [])]


def document() -> dict:
    """The templates and the whole catalogue, aliases included, as plain data.

    The whole catalogue rather than only what the templates use: finding
    duplicates needs every exercise's aliases.
    """
    return {
        "catalogue": [
            {
                "name": name,
                "category": category,
                "modality": modality,
                "muscles": muscles,
                "aliases": ALIASES.get(name, []),
            }
            for name, category, modality, muscles in CATALOGUE
        ],
        "templates": [
            {
                "id": t["id"],
                "name": t["name"],
                "about": t["about"],
                "exercises": [
                    {"exercise": name, "sets": sets, "reps": reps}
                    for name, sets, reps in t["exercises"]
                ],
            }
            for t in TEMPLATES
        ],
    }


def template(template_id: str) -> dict | None:
    return next((t for t in document()["templates"] if t["id"] == template_id), None)


@dataclass(frozen=True)
class Candidate:
    """An existing exercise, with what counts when choosing between names."""

    id: int
    name: str
    sets: int
    imported: int
    rank: int

    @property
    def key(self) -> tuple:
        # Imported history first, then the most working sets, then the
        # earlier name in the alias list, then the older row.
        return (self.imported > 0, self.sets, -self.rank, -self.id)


def _existing(session: Session) -> list[tuple[int, str, int, int]]:
    """Every exercise with its working sets, and how many of them were imported."""
    return [
        (row.id, row.name, row.sets, row.imported)
        for row in session.execute(
            select(
                Exercise.id,
                Exercise.name,
                func.count(SetEntry.id).label("sets"),
                func.count(case((Workout.source != "manual", SetEntry.id))).label("imported"),
            )
            .outerjoin(
                SetEntry,
                and_(SetEntry.exercise_id == Exercise.id, SetEntry.is_warmup.is_(False)),
            )
            .outerjoin(Workout, Workout.id == SetEntry.workout_id)
            .group_by(Exercise.id, Exercise.name)
        )
    ]


def _candidates(existing: list[tuple[int, str, int, int]], exercise: str) -> list[Candidate]:
    """The existing exercises that go by this catalogue exercise's names, best first."""
    rank: dict[str, int] = {}
    for i, name in enumerate(names_for(exercise)):
        rank.setdefault(normalise(name), i)
    found = [
        Candidate(id_, name, sets, imported, rank[normalise(name)])
        for id_, name, sets, imported in existing
        if normalise(name) in rank
    ]
    return sorted(found, key=lambda c: c.key, reverse=True)


def resolve(session: Session, exercise: str) -> Exercise | None:
    """The existing exercise a template's exercise means, if there is one.

    Any exercise whose name matches the template's name or one of its aliases
    counts. When several do, as when an earlier workout created "Bench Press"
    beside an imported "Flat Barbell Bench Press", the imported one wins, then
    the one with the most working sets, since that is where the history is.
    """
    found = _candidates(_existing(session), exercise)
    return session.get(Exercise, found[0].id) if found else None


@dataclass(frozen=True)
class Merge:
    keep: Candidate
    drop: list[Candidate]

    def as_dict(self) -> dict:
        def one(c: Candidate) -> dict:
            return {"id": c.id, "name": c.name, "sets": c.sets}

        return {"keep": one(self.keep), "drop": [one(c) for c in self.drop]}


def duplicates(session: Session) -> list[Merge]:
    """Catalogue exercises that exist under more than one of their names."""
    existing = _existing(session)
    merges, claimed = [], set()
    for name, *_ in CATALOGUE:
        found = [c for c in _candidates(existing, name) if c.id not in claimed]
        if len(found) > 1:
            merges.append(Merge(found[0], found[1:]))
            claimed.update(c.id for c in found)
    return merges


def merge_duplicates(session: Session) -> list[Merge]:
    """Fold each duplicate into the exercise `duplicates` chose to keep.

    Its sets move across and it is deleted. The kept exercise keeps its own
    name, category and muscles, taking the duplicate's only where it has none,
    so an import that brought no muscle mapping gains the catalogue's. One
    transaction: a failure part-way leaves everything as it was.
    """
    merges = duplicates(session)
    for merge in merges:
        keep = merge.keep.id
        for drop in (c.id for c in merge.drop):
            params = {"keep": keep, "drop": drop}
            session.execute(
                text("UPDATE sets SET exercise_id = :keep WHERE exercise_id = :drop"), params
            )
            session.execute(
                text(
                    """
                    INSERT INTO exercise_muscles (exercise_id, muscle, is_primary)
                    SELECT :keep, muscle, is_primary FROM exercise_muscles
                    WHERE exercise_id = :drop
                      AND NOT EXISTS (SELECT 1 FROM exercise_muscles WHERE exercise_id = :keep)
                    """
                ),
                params,
            )
            session.execute(
                text(
                    """
                    UPDATE exercises
                    SET category = (SELECT category FROM exercises WHERE id = :drop)
                    WHERE id = :keep AND category IS NULL
                    """
                ),
                params,
            )
            session.execute(text("DELETE FROM exercise_muscles WHERE exercise_id = :drop"), params)
            session.execute(text("DELETE FROM exercises WHERE id = :drop"), params)
    session.commit()
    session.expire_all()
    return merges


def ensure_exercises(session: Session, chosen: dict) -> None:
    """Create the template's exercises that exist under none of their names.

    An existing exercise is left exactly as it is, even if its muscles differ
    from the catalogue's: it may have come from an import, and history already
    hangs off it.
    """
    for entry in chosen["exercises"]:
        if resolve(session, entry["exercise"]) is not None:
            continue
        name, category, modality, muscles = _CATALOGUE[entry["exercise"].lower()]
        session.add(
            Exercise(name=name, category=category, modality=modality, primary_muscles=muscles)
        )
        session.flush()
    session.commit()


def start(session: Session, template_id: str, day: dt.date) -> dict | None:
    chosen = template(template_id)
    if chosen is None:
        return None
    ensure_exercises(session, chosen)
    settings_store.put(session, ACTIVE, {"template": template_id, "day": day.isoformat()})
    return plan(session, day)


def finish(session: Session) -> None:
    settings_store.delete(session, ACTIVE)


def plan(session: Session, day: dt.date) -> dict | None:
    """Today's plan, if one was started today, with progress through it.

    Each entry names the exercise as it exists in the database, which may be
    an alias of the template's name; `planned` keeps the template's. `done`
    counts working sets of it logged on the day, from any workout — logging
    it by hand or through the coach counts too. `last_weight_kg` is the most
    recent working weight, including today's, so the next set starts where
    the last one left off.
    """
    active = settings_store.get(session, ACTIVE) or {}
    if active.get("day") != day.isoformat():
        return None
    chosen = template(active.get("template", ""))
    if chosen is None:
        return None

    entries = []
    for entry in chosen["exercises"]:
        exercise = resolve(session, entry["exercise"])
        done, last = 0, None
        if exercise is not None:
            done = session.scalar(
                select(func.count())
                .select_from(SetEntry)
                .join(Workout, Workout.id == SetEntry.workout_id)
                .where(
                    SetEntry.exercise_id == exercise.id,
                    SetEntry.is_warmup.is_(False),
                    Workout.performed_on == day,
                )
            ) or 0
            last = session.scalar(
                select(SetEntry.weight_kg)
                .join(Workout, Workout.id == SetEntry.workout_id)
                .where(
                    SetEntry.exercise_id == exercise.id,
                    SetEntry.is_warmup.is_(False),
                    SetEntry.weight_kg.is_not(None),
                )
                .order_by(Workout.performed_on.desc(), SetEntry.id.desc())
                .limit(1)
            )
        entries.append(
            {
                "exercise": exercise.name if exercise is not None else entry["exercise"],
                "planned": entry["exercise"],
                "exercise_id": exercise.id if exercise is not None else None,
                "sets": entry["sets"],
                "reps": entry["reps"],
                "done": done,
                "last_weight_kg": float(last) if last is not None else None,
            }
        )
    return {"template": chosen, "day": day.isoformat(), "entries": entries}
